mapper/software_management/src/software.rs

```diff
@@ -0,0 +1,133 @@
+}
+
+pub type SoftwareList = Vec<SoftwareModule>;
+
+#[derive(Debug, Clone, Deserialize, Serialize, PartialEq)]
+#[serde(untagged)]
+pub enum SoftwareOperation {
+    // A request for the current software list
+    CurrentSoftwareList,
+
+    // A sequence of updates to be applied
+    SoftwareUpdates { updates: Vec<SoftwareUpdate> },
+
+    // The desired software list
+    DesiredSoftwareList { modules: Vec<SoftwareModule> },
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize, PartialEq)]
+#[serde(tag = "action")]
+pub enum SoftwareUpdate {
+    #[serde(rename = "install")]
+    Install {
+        #[serde(flatten)]
+        module: SoftwareModule,
+    },
+
+    #[serde(rename = "uninstall")]
+    UnInstall {
+        #[serde(flatten)]
+        module: SoftwareModule,
+    },
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize, PartialEq)]
+pub enum SoftwareOperationStatus {
+    SoftwareUpdates { updates: Vec<SoftwareUpdateStatus> },
+    DesiredSoftwareList { updates: Vec<SoftwareUpdateStatus> },
+    CurrentSoftwareList { modules: Vec<SoftwareModule> },
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize, PartialEq)]
+pub struct SoftwareUpdateStatus {
+    pub update: SoftwareUpdate,
+    pub status: UpdateStatus,
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize, PartialEq)]
+pub enum UpdateStatus {
+    Scheduled,
+    Success,
+    Error { reason: SoftwareError },
+    Cancelled,
+}
+
+#[derive(thiserror::Error, Debug, Clone, Deserialize, Serialize, PartialEq)]
+pub enum SoftwareError {
+    #[error("JSON parse error: {reason:?}")]
+    ParseError { reason: String },
+
+    #[error("Unknown software type: {software_type:?}")]
+    UnknownSoftwareType { software_type: SoftwareType },
+
+    #[error("Unknown {software_type:?} module: {name:?}")]
+    UnknownModule {
+        software_type: SoftwareType,
+        name: SoftwareName,
+    },
+
+    #[error("Unknown {software_type:?} version: {name:?} - {version:?}")]
+    UnknownVersion {
+        software_type: SoftwareType,
+        name: SoftwareName,
+        version: SoftwareVersion,
+    },
+
+    #[error("Unexpected module type: actual: {actual_type:?}, expected: {expected_type:?}")]
+    WrongModuleType {
+        actual_type: SoftwareType,
+        expected_type: SoftwareType,
+    },
+
+    #[error("Plugin error for {software_type:?}, reason: {reason:?}")]
+    PluginError {
+        software_type: SoftwareType,
+        reason: String,
+    },
+
+    #[error("Fail to install {module:?}")]
+    InstallError {
+        module: SoftwareModule,
+        reason: String,
+    },
+
+    #[error("Fail to uninstall {module:?}")]
+    UnInstallError {
+        module: SoftwareModule,
+        reason: String,
+    },
+}
+
+impl From<serde_json::Error> for SoftwareError {
+    fn from(err: serde_json::Error) -> Self {
+        SoftwareError::ParseError {
+            reason: format!("{}", err),
+        }
+    }
+}
+
+impl SoftwareUpdate {
+    pub fn module(&self) -> &SoftwareModule {
```

Comment:
Didn't understand the purpose of this function.

Reply:
This is a getter returning the module to be updated.